crates/sparrow-scheduler/src/pipeline.rs

```diff
@@ -0,0 +1,109 @@
+use sparrow_arrow::Batch;
+
+use crate::queue::Queue;
+use crate::{Partition, TaskRef, Worker};
+
+/// A push-based interface used by the scheduler to drive query execution
```

Comment:
I am not sure I am getting the relationships between Pipeline, partition, partition index, children and batch. 

Maybe we should have a comment that show an example or at least goes into how these things hook up to each other.

Reply:
Yeah, I'll probably try to add one to the top level. But to develop it, I'll try to answer here and then maybe you, @jordanrfrazier and I can discuss?

1. The entire query plan consists of a bunch of "transforms" (lookup request, lookup response, project, select, etc.)
2. We divide those "transforms" up into "pipeline plans" that are a single input -> sequence of transforms -> single output.
3. For execution, we create a bunch of "pipelines" that manage the data for a specific part of execution. This includes "transform pipelines" that run transforms, but also "repartition", "scan", etc. (eg., everything that executes is a pipeline, and it may be a superset of the "pipeline plans").
4. Each pipeline manages the state for *all* partitions that it may execute on. This is intentional, since it allows things like the "repartition" pipeline to accept data from any producing / input partition and store it in queues for each of the downstream partitions.
5. The above means that a pipeline has "input partitions" (the partitioning of the producer) and "processing / output partitions" (the partitioning of the actual pipeline and the output it produces).
6. A pipeline has 0 or more inputs (eg., a merge operation has two inputs, many things have 1 input, etc.). Thus, inputs are associated with the "input partition" (the partition that produced the input) and the "input index" (which of the inputs to the pipeline it is being sent to.
7. Children I don't think matters (it's basically how pipelines are connected and/or which transforms are connected together within a pipeline).
8. Batches are data that flows between pipelines.

Reply:
1. Not sure yet. I would expect that there is a step in the physical plan for applying aggregations, and then the individual aggregations are expressions within that step. Aggregations are somewhat TBD, but my thinking is that we'd have a step that says "do aggregations with window W" (so the cardinality is determined by the step) and then all the aggregations with that windowing are part of that step. Transforms and steps and pipelines probably merit some discussion.
2. Roughly. Not all pipelines at execution time correspond to "pipeline plans" in the proto.
3. `Scan` is a pipeline. Transform is just a class for "common" steps that behave a certain way, so we don't need to write them as separate pipelines, and transforms allow chaining a few them into a single pipeline. But special things (input, output, repartition, etc.) are pipelines on their own.
4. A bit of both -- the `with_key` function is the logic of changing the `key` and then it applies the `reparatition` pipeline to actually change partitioning and move data around.
5.  A pipeline can have different input/output partitions if needed, but I don't know that we'll have pipelines with that property often.
6. I think `input partition` is the partitioning from the producer. An interesting question is if we have multiple inputs do we require all of them to have the same partitioning, or do we allow it to be different. I think this could also be per-pipeline. Eg., I could see a case of the `merge` pipeline *not* having that requirement, and mixing some "repartition" logic into the merge operation.

Reply:
> 
>     1. The entire query plan consists of a bunch of "transforms" (lookup request, lookup response, project, select, etc.)
> 
>     2. We divide those "transforms" up into "pipeline plans" that are a single input -> sequence of transforms -> single output.
> 

  Do you mean we "group" rather than "divide" here? Are transforms the leafs of an execution (indivisible computation like an `add`) or are they decomposable? My reading of this let me to understand that transforms are the actions and pipelines compose them together. 

 What really throws me off here is that we "divide" transforms into pipeline plans that consists of a sequence of transforms again? 

  Maybe my extrapolation from the English words is getting in the way here. I would expect a pipe to be an individual element. Pipelines to compose these pipes together in some way. 
I am not seeing this connection here with transformers, pipeline plans, pipelines and the like. 

> 
>     4. Each pipeline manages the state for _all_ partitions that it may execute on. This is intentional, since it allows things like the "repartition" pipeline to accept data from any producing / input partition and store it in queues for each of the downstream partitions.

 So partitions have to do with what flows into the pipe, i.e., data. What is the relationship with a Batch? Is a batch a segment of data from 1 partition only, or, can it have data from multiple partitions?  


> 
>     5. The above means that a pipeline has "input partitions" (the partitioning of the producer) and "processing / output partitions" (the partitioning of the actual pipeline and the output it produces).

  These partitions are something that is checked/verified so that outputs from one pipeline "match" inputs to the succeeding pipeline? 

> 
>     6. A pipeline has 0 or more inputs (eg., a merge operation has two inputs, many things have 1 input, etc.). Thus, inputs are associated with the "input partition" (the partition that produced the input) and the "input index" (which of the inputs to the pipeline it is being sent to.

  I think I am missing something here. We have multiple inputs so there is an index for each input, I assume. Now each input has partitions so there is another index within the input to designate a partition, e.g., 1:3 for input 1 and partition 3 of input 1? Is this the setup? Are partitions internal components of inputs? 

> 
>     7. Children I don't think matters (it's basically how pipelines are connected and/or which transforms are connected together within a pipeline).
>

  Well I would expect children to matter a little since if they specify connections we need to ensure that the connection have right input+partition for the next step in the pipeline, no? 
 
>     8. Batches are data that flows between pipelines.

  But what is the relationship here. Is batch a segment of a stream of data coming from 1 partition connected to 1 input *only*?